Fix no-tls1_2

levitte opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] tests are added or updated
- [x] CLA is signed

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->

levitte committed:
70-test_sslmessages.t: Don't check EXT_SIG_ALGS if TLS 1.2 is disabled

levitte committed:
70-test_sslvertol.t: Make sure to check a max TLS version that matches configuration

levitte committed:
80-test_ssl_new.t: Make 19-mac-then-encrypt.conf work without TLSv1.2

levitte commented:
Note that I haven't checked how this backports to 1.1.0 and 1.0.2, or if that's even needed.

levitte commented:
Quick check that 1.1.0 is fine

kroeckx commented:
This seems to break the enable-tls1_3 target

richsalz reviewed (approved):
and also 1.1.0?

dot-asm commented:
Side note. I'd love to see criteria for which no-options should be supported. I mean I see lesser value in supporting arbitrary no-option, as it's nothing but ever increasing support burden. Or in other words it would be appropriate to minimize amount of no-options, in other words limit it to necessary minimum, hence the question about criteria [for defined the said minimum]. Why is it side note specifically here? Because I'd argue that no-tls1_2 should not be a supported option, because a) it makes build less secure, b) alleged goal can be achieved programmatically even if 1.2 support is compiled. This is general comment...

vdukhovni commented:
I'm inclined to agree that no-tls1_2 need not be implemented at this time. IIRC removing a TLS protocol at configure time removes all lower protocols also, and it'll be a while before TLS-1.3-only is realistic.

richsalz commented:
This might be useful:
```
; git grep OPENSSL_NO_ | grep ifn*def | sed -e 's/.*ifn*def//' | sort | uniq -c | sort -n
```

dot-asm commented:
> This might be useful:

Right, and what I'm talking about that they don't have to be that many... But this getting out of topic. I mean how many are there is out of scope, what I'm trying to say is that it's not given that there should be no-tls1_2, because I find it figure out reason why should there be one... Once again, this is side note, i.e. not something that needs to be acted upon in context of this request.

dot-asm commented:
> I find it figure out reason why should there be one

"I find it *hard* to figure out the reason" naturally...

mattcaswell commented:
> This might be useful:

All supported options should be the ones listed in INSTALL/Configure. There are a few that exist in the codebase that can't be set via Configure (e.g. no-rsa).

mattcaswell reviewed (changes requested):
The `enable-tls1_3` target needs to be fixed before this can be merged.

levitte commented:
Ah, good point.  Will do

levitte commented:
@mattcaswell, a moment of your time, please?  In `70-test_sslvertol.t`, what's the actual difference between test 1 and test 2?  As far as I can see, they do *exactly* the same test.  With those, I tried removing this line, and that seems to work, at least preliminarly:

    $proxy->clientflags("-no_tls1_3");

Also, I can't make sense of test 3.  If I remove that same line (above) from that test, it fails, when when I check the last record, it has version number 772 (which stands for TLS 1.3, should anyone else wonder), and I frankly don't quite understand what's going on under the hood...

This is with `enable-tls1_3` configured.

levitte commented:
Note: when I do a default configuration or configure with `no-tls1_2`, `70-test_sslvertol.t` passes beautifully with that line (repeated here) removed...

    $proxy->clientflags("-no_tls1_3");

mattcaswell commented:
> @mattcaswell, a moment of your time, please? In 70-test_sslvertol.t, what's the actual difference between test 1 and test 2? As far as I can see, they do exactly the same test.

Ahh. Actually I think that is a bug in the test. As you can see from the comment Test 2 is supposed to be asking for TLS1.3. But `$clientversion` is still set to TLS1.4. It should be set to TLS1.3 for that test.

> With those, I tried removing this line, and that seems to work, at least preliminarly:

Hmm. The purpose of the "-no_tls1_3" is to stop the client from creating the supported_versions extension. It won't create that anyway if tls1.3 is disabled, but will if it is enabled. The whole point of this test is to check what happens without that extensions being present (see the comment on line 37)

> Also, I can't make sense of test 3. If I remove that same line (above) from that test, it fails, when when I check the last record, it has version number 772 (which stands for TLS 1.3, should anyone else wonder), and I frankly don't quite understand what's going on under the hood...

The supported_versions extension is kicking in and therefore ignoring the client_version setting.

levitte commented:
OK, I've made some changes to `70-test_sslvertol.t` that should cover the cases that we want to test for and that tolerates different configurations.

mattcaswell commented on test/recipes/80-test_ssl_new.t:
```diff
@@ -78,7 +78,7 @@ my %skip = (
   "16-dtls-certstatus.conf" => $no_dtls || $no_ocsp,
   "18-dtls-renegotiate.conf" => $no_dtls,
-  "19-mac-then-encrypt.conf" => $no_pre_tls1_3
+  "19-mac-then-encrypt.conf" => $no_pre_tls1_3 || disabled("tls1_2")
```
Why is this test TLSv1.2 specific?

levitte replied:
I haven't quite understood why, I just know that it fails if TLSv1.2 is disabled (`no-tls1_2`).  Maybe you have an explanation?

mattcaswell replied:
Take a look in 19-mac-then-encrypt.conf.in. My guess would be that the final 3 sub-tests use TLSv1.2 only ciphersuites. It would probably be better just to disable those three sub-tests in the no-tls1_2 case. If you do that then you'll probably also need to add this test to `%conf_dependent_tests` in this file.

levitte replied:
Okie, looking more closely at this.

mattcaswell commented on test/recipes/70-test_sslvertol.t:
```diff
@@ -34,33 +34,71 @@ my $proxy = TLSProxy::Proxy->new(
+ok((note("Record version received: ".$record->version()),
+    TLSProxy::Message->success())
+   && $record->version() == tls_version_below(TLSProxy::Record::VERS_TLS_1_4),
+   "Version tolerance test, anything up to and including TLS 1.3");
```
This test is quite confusing. By removing the "-no_tls1_3" flag you have effectively made the client send the supported_versions extension which means the server should ignore the client_version value (which is normally set to TLS1.2, but you are setting to TLS1.3 in this test). However if TLSv1.3 is disabled then no supported_versions extension is sent and you are testing something different. It effectively becomes (almost) the same test as Test 1. If you think this test is necessary then I think we need to be clearer about what it is actually intended to do. If we have it then the comment on line 55 is no longer correct.

levitte replied:
`supported_versions` had escaped me.  I'll simply remove this test.

As for line 55, I'm not sure how that's relevant.  That only tells which protocols are available per OpenSSL configuration, nothing else.  The tests may then apply their own conditions.  Basically, I needed line 55 so it was clear to me exactly what protocol numbers I can expect at any time.

mattcaswell replied:
Either you or I have read the line number wrong! I meant this one:

````
#This file does tests without the supported_versions extension.
````

levitte committed:
70-test_sslvertol.t: skip test 1 and 2 if too few protocols are enabled

These tests depend on there being at least one protocol version below
TLSv1.3 enabled.

levitte commented:
I made changes to sslvertol that I hope correspond to what was said

mattcaswell commented on test/recipes/70-test_sslvertol.t:
```diff
@@ -34,33 +34,69 @@ my $proxy = TLSProxy::Proxy->new(
+#Test 1: Asking for TLS1.4 should pass and negotiate the maximum
+#available TLS version according to configuration
 my $client_version = TLSProxy::Record::VERS_TLS_1_4;
-$proxy->clientflags("-no_tls1_3");
```
We are trying to test what happens without the supported_versions extension which this flag gives us - so this needs to go back in.

levitte replied:
Okie

mattcaswell commented on test/recipes/70-test_sslvertol.t:
```diff
@@ -34,33 +34,69 @@ my $proxy = TLSProxy::Proxy->new(
+ok((note("Record version received: ".
+         (defined $record ? $record->version() : "none")),
+    TLSProxy::Message->success())
+   && $record->version() == tls_version_below(TLSProxy::Record::VERS_TLS_1_4),
```
Without supported_versions the max negotiated version should be TLSv1.2...so the above needs to say VERS_TLS_1_3

levitte replied:
Will do

levitte commented:
Cleaned up the mess of commits and rebased on fresh master

mattcaswell commented on test/recipes/70-test_sslvertol.t:
```diff
@@ -34,33 +34,75 @@ my $proxy = TLSProxy::Proxy->new(
+       "Version tolerance test, max version but not TLS 1.3");
+}
+
+#Test 3: Testing something below SSLv3 should fail.  We must disabled TLS 1.3
```
typo: disable

levitte replied:
Done

mattcaswell reviewed (approved):
Approved subject to fixing the typo

levitte commented:
Merged.
